pkg/tools/dnsutils/cache/handler.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2022 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cache stores successful requests to DNS server
+package cache
+
+import (
+	"context"
+	"time"
+
+	"github.com/miekg/dns"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/dnsutils"
+	"github.com/networkservicemesh/sdk/pkg/tools/dnsutils/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+)
+
+type dnsCacheHandler struct {
+	cache         *Map
+	lastTTLUpdate time.Time
+}
+
+func (h *dnsCacheHandler) ServeDNS(ctx context.Context, rw dns.ResponseWriter, m *dns.Msg) {
+	h.updateTTL()
+	if v, ok := h.cache.Load(m.Question[0].Name); ok {
+		if validateMsg(v) {
+			v.Id = m.Id
+			if err := rw.WriteMsg(v); err != nil {
+				log.FromContext(ctx).Warnf("got an error during write the message: %v", err.Error())
+				dns.HandleFailed(rw, v)
+				return
+			}
+			return
+		}
+
+		h.cache.Delete(m.Question[0].Name)
+	}
+
+	wrapper := responseWriterWrapper{
+		ResponseWriter: rw,
+		cache:          h.cache,
+	}
+
+	next.Handler(ctx).ServeDNS(ctx, &wrapper, m)
+}
+
+func (h *dnsCacheHandler) updateTTL() {
+	now := time.Now()
+
+	diff := uint32(now.Sub(h.lastTTLUpdate).Seconds())
+	if diff == 0 {
```

Comment:
Can we face uint overflow here?

Reply:
I don't think so. Max value of uint32 is 2^32
2^32 / 60 / 60 / 24 / 365 = 136 years
If we wait for about 136 years maybe it will overflow

On first request we might get some problems because we don't initialize `lastTTLUpdate` anywhere.

Reply:
Added `lastTTLUpdate` initialization